File: src/test/java/cube8540/oauth/authentication/resource/domain/SecuredResourceTest.java
package cube8540.oauth.authentication.resource.domain;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static cube8540.oauth.authentication.resource.domain.SecuredResourceTestHelper.ACCESSIBLE_AUTHORITY;
import static cube8540.oauth.authentication.resource.domain.SecuredResourceTestHelper.CHANGE_RESOURCE;
import static cube8540.oauth.authentication.resource.domain.SecuredResourceTestHelper.CHANGE_RESOURCE_METHOD;
import static cube8540.oauth.authentication.resource.domain.SecuredResourceTestHelper.RAW_AUTHORITY_CODE;
import static cube8540.oauth.authentication.resource.domain.SecuredResourceTestHelper.RESOURCE;
import static cube8540.oauth.authentication.resource.domain.SecuredResourceTestHelper.RESOURCE_ID;
import static cube8540.oauth.authentication.resource.domain.SecuredResourceTestHelper.RESOURCE_METHOD;
import static cube8540.oauth.authentication.resource.domain.SecuredResourceTestHelper.makeErrorValidatorFactory;
import static cube8540.oauth.authentication.resource.domain.SecuredResourceTestHelper.makePassValidatorFactory;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

@DisplayName("보호 자원 테스트")
class SecuredResourceTest {

    @Test
    @DisplayName("자원 정보 변경")
    void changeResourceInfo() {
        SecuredResource resource = new SecuredResource(RESOURCE_ID, RESOURCE, RESOURCE_METHOD);

        resource.changeResourceInfo(CHANGE_RESOURCE, CHANGE_RESOURCE_METHOD);
        assertEquals(CHANGE_RESOURCE, resource.getResource());
        assertEquals(CHANGE_RESOURCE_METHOD, resource.getMethod());
    }

    @Test
    @DisplayName("접근 권한 추가")
    void addAuthority() {
        SecuredResource resource = new SecuredResource(RESOURCE_ID, RESOURCE, RESOURCE_METHOD);

        resource.addAuthority(RAW_AUTHORITY_CODE);
        assertTrue(resource.getAuthorities().contains(ACCESSIBLE_AUTHORITY));
    }

    @Test
    @DisplayName("접근 권한 삭제")
    void removeAuthority() {
        SecuredResource resource = new SecuredResource(RESOURCE_ID, RESOURCE, RESOURCE_METHOD);
        resource.addAuthority(RAW_AUTHORITY_CODE);

        resource.removeAuthority(RAW_AUTHORITY_CODE);
        assertFalse(resource.getAuthorities().contains(ACCESSIBLE_AUTHORITY));
    }

    @Test
    @DisplayName("유효 하지 않은 정보가 저장 되어 있을시")
    void resourceDataInvalid() {
        SecuredResource resource = new SecuredResource(RESOURCE_ID, RESOURCE, RESOURCE_METHOD);

        SecuredResourceValidatorFactory factory = makeErrorValidatorFactory(resource);

        assertThrows(ResourceInvalidException.class, () -> resource.validation(factory));
    }

    @Test
    @DisplayName("모든 데이터가 유효할시")
    void resourceDataAllowed() {
        SecuredResource resource = new SecuredResource(RESOURCE_ID, RESOURCE, RESOURCE_METHOD);

        SecuredResourceValidatorFactory factory = makePassValidatorFactory(resource);

        assertDoesNotThrow(() -> resource.validation(factory));
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/resource/domain/SecuredResourceTest.kt
package cube8540.oauth.authentication.resource.domain

import cube8540.validator.core.ValidationError
import cube8540.validator.core.ValidationRule
import cube8540.validator.core.Validator
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.*
import org.junit.jupiter.api.*
import java.net.URI

class SecuredResourceTest {

    @Nested
    inner class ValidationTest {
        private val resourceValidatorFactory: SecuredResourceValidatorFactory = mockk()
        private val resource = SecuredResource(SecuredResourceId("resourceId"), URI.create("http://localhost"), ResourceMethod.ALL)

        @Test
        fun `resource data is invalid`() {
            val rule = mockk<ValidationRule<SecuredResource>> {
                every { isValid(resource) } returns false
                every { error() } returns ValidationError("resourceId", "test")
            }
            every { resourceValidatorFactory.createValidator(resource) } returns Validator.of(resource).registerRule(rule)

            val thrown = catchThrowable { resource.validation(resourceValidatorFactory) }
            assertThat(thrown).isInstanceOf(ResourceInvalidException::class.java)
        }

        @Test
        fun `resource data is allowed`() {
            val rule = mockk<ValidationRule<SecuredResource>> {
                every { isValid(resource) } returns true
            }
            every { resourceValidatorFactory.createValidator(resource) } returns Validator.of(resource).registerRule(rule)

            assertThatCode { resource.validation(resourceValidatorFactory) }.doesNotThrowAnyException()
        }
    }

    @Nested
    inner class ChangeResourceInfoTest {
        private val resource = SecuredResource(SecuredResourceId("resourceId"), URI.create("http://localhost"), ResourceMethod.ALL)

        @Test
        fun `change resource info`() {
            resource.changeResourceInfo(URI.create("http://change-host"), ResourceMethod.POST)

            assertThat(resource.resource).isEqualTo(URI.create("http://change-host"))
            assertThat(resource.method).isEqualTo(ResourceMethod.POST)
        }
    }

    @Nested
    @TestMethodOrder(value = MethodOrderer.OrderAnnotation::class)
    inner class ChangeAuthorityTest {
        private val resource = SecuredResource(SecuredResourceId("resourceId"), URI.create("http://localhost"), ResourceMethod.ALL)

        @Test
        @Order(1)
        fun `add authority`() {
            resource.addAuthority("ADD_AUTHORITY")

            assertThat(resource.authorities).contains(AccessibleAuthority("ADD_AUTHORITY"))
        }

        @Test
        @Order(2)
        fun `remove authority`() {
            resource.removeAuthority("ADD_AUTHORITY")

            assertThat(resource.authorities).doesNotContain(AccessibleAuthority("ADD_AUTHORITY"))
        }
    }
}